File: Python/Problems/Problems/problem1296.py
from .problem import Problem


class Problem1296(Problem):
    def test(self):
        nums = [3, 2, 1, 2, 3, 4, 3, 4, 5, 9, 10, 11]
        k = 3

        expected = True

        result = self.isPossibleDivide(nums, k)

        return expected == result

    def isPossibleDivide(self, nums, k):
        n = len(nums)
        if n % k != 0:
            return False
        freq = {}
        for num in nums:
            if (num in freq):
                freq[num] += 1
            else:
                freq[num] = 1
        sorted = list(freq.keys())
        sorted.sort()

        for key in sorted:
            f = freq[key]
            if f == 0:
                continue

            for i in range(k):
                next = key + i
                if next not in freq or freq[next] < f:
                    return False
                freq[next] -= f


        return True

File: C++/Problems/Problems/problem_1296.cpp
#include "problem_1296.h"

#include <map>

bool problem_1296::test() {
    std::vector nums { 3, 2, 1, 2, 3, 4, 3, 4, 5, 9, 10, 11 };
    auto k = 3;

    auto expected = true;

    auto result = isPossibleDivide(nums, k);

    return expected == result;
}

bool problem_1296::isPossibleDivide(std::vector<int> &nums, int k) {
    int n = nums.size();
    if (n % k != 0)
        return false;
    std::map<int, int> freq;
    for (auto num : nums)
    {
        freq[num]++;
    }

    for (auto pair : freq)
    {
        int f = freq[pair.first];
        if (f == 0) continue;

        for (int i = 0; i < k; i++)
        {
            int next = pair.first + i;
            if (!freq.contains(next) || freq[next] < f)
                return false;
            freq[next] -= f;
        }
    }

    return true;
}
